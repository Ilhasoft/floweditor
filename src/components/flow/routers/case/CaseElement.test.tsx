import CaseElement, { CaseElementProps } from 'components/flow/routers/case/CaseElement';
import { Operators } from 'config/interfaces';
import * as React from 'react';
import {
  fireEvent,
  render,
  fireTembaSelect,
  fireUnnnicSelect,
  fireUnnnicInputChangeText,
  getUnnnicInputValue
} from 'test/utils';
import { createUUID } from 'utils';

const caseUUID = createUUID();

const caseProps: CaseElementProps = {
  name: `case_${caseUUID}`,
  kase: {
    uuid: caseUUID,
    type: Operators.has_any_word,
    arguments: [''],
    category_uuid: createUUID()
  },
  categoryName: null,
  onRemove: jest.fn(),
  onChange: jest.fn()
};

const selectOperator = (getByTestId: any, operator: Operators) => {
  fireUnnnicSelect(getByTestId('temba_select_operator'), { value: [{ type: operator }] }, 'value');
};

describe(CaseElement.name, () => {
  describe('render', () => {
    it('should render empty case', () => {
      const { baseElement } = render(<CaseElement {...caseProps} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('renders no argument rules', () => {
      const kase = {
        uuid: caseUUID,
        type: Operators.has_number,
        category_uuid: createUUID()
      };

      const { baseElement } = render(<CaseElement {...caseProps} kase={kase} />);
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('operator changes', () => {
    // we need a full renders because we update our category reference
    it('should handle updates', () => {
      const { baseElement, getByTestId } = render(<CaseElement {...caseProps} />);

      selectOperator(getByTestId, Operators.has_phone);
      expect(baseElement).toMatchSnapshot();
    });

    it('should should set arguments for numeric range', () => {
      const { baseElement, getByTestId } = render(<CaseElement {...caseProps} />);
      selectOperator(getByTestId, Operators.has_number_between);
      expect(baseElement).toMatchSnapshot();
    });

    it('shouldnt update exit if it has been edited', () => {
      const { baseElement, queryByDisplayValue, getByTestId } = render(
        <CaseElement {...caseProps} />
      );

      // make us a has phone so we can look up our category by value
      selectOperator(getByTestId, Operators.has_phone);

      // update our category to a user supplied value
      const category = getByTestId('Exit Name');
      fireUnnnicInputChangeText(category, 'My Exit Name');

      // now swithc our type to force a category change
      selectOperator(getByTestId, Operators.has_number);

      // we shouldn't have updated our category
      expect(queryByDisplayValue('My Exit Name')).not.toBeNull();
      expect(queryByDisplayValue('Has Number')).toBeNull();
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('handles removes', () => {
      const onRemove = jest.fn();
      const { getByTestId } = render(<CaseElement {...caseProps} onRemove={onRemove} />);

      fireEvent.click(getByTestId(`remove-case-${caseProps.kase.uuid}`));
      expect(onRemove).toHaveBeenCalled();
    });

    it('handles argument change', () => {
      const onRemove = jest.fn();
      const { baseElement, getAllByTestId } = render(
        <CaseElement {...caseProps} onRemove={onRemove} />
      );
      const args = getAllByTestId('arguments');
      fireUnnnicInputChangeText(args[0], 'Purple, p');
      expect(getUnnnicInputValue(args[0])).toEqual('Purple, p');
      expect(baseElement).toMatchSnapshot();
    });

    it('handles multiple argument change', () => {
      const onRemove = jest.fn();
      const { baseElement, getByTestId, getAllByTestId } = render(
        <CaseElement {...caseProps} onRemove={onRemove} />
      );

      selectOperator(getByTestId, Operators.has_number_between);

      const args = getAllByTestId('arguments');

      fireUnnnicInputChangeText(args[0], '1');
      fireUnnnicInputChangeText(args[1], '100');
      expect(getUnnnicInputValue(args[0])).toEqual('1');
      expect(getUnnnicInputValue(args[1])).toEqual('100');
      expect(baseElement).toMatchSnapshot();
    });
  });
});
