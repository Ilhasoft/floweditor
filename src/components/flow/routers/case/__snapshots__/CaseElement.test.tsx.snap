// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseElement operator changes should handle updates 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_phone"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_phone\\",\\"verboseName\\":\\"has a phone number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Phone\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value="Has Phone"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement operator changes should should set arguments for numeric range 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number_between"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number_between\\",\\"verboseName\\":\\"has a number between\\",\\"operands\\":2}]"
            />
          </div>
        </div>
        <div
          class="multi_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                type="text"
                value=""
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <span
            class="divider"
            data-draggable="true"
          >
            and
          </span>
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                type="text"
                value=""
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value=""
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement operator changes shouldnt update exit if it has been edited 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number\\",\\"verboseName\\":\\"has a number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Number\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value="My Exit Name"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement render renders no argument rules 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number\\",\\"verboseName\\":\\"has a number\\",\\"operands\\":0,\\"categoryName\\":\\"Has Number\\"}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        />
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value="Has Number"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement render should render empty case 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_any_word"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_any_word\\",\\"verboseName\\":\\"has any of the words\\",\\"operands\\":1,\\"visibility\\":[\\"message\\",\\"message_offline\\"]}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                placeholder=""
                type="text"
                value=""
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value=""
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement update handles argument change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_any_word"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_any_word\\",\\"verboseName\\":\\"has any of the words\\",\\"operands\\":1,\\"visibility\\":[\\"message\\",\\"message_offline\\"]}]"
            />
          </div>
        </div>
        <div
          class="single_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                placeholder=""
                type="text"
                value="Purple, p"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value="Purple"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CaseElement update handles multiple argument change 1`] = `
<body>
  <div>
    <div
      class="ele group"
    >
      <div
        class="kase has_number_between"
        data-draggable="true"
      >
        <span
          class="fe-chevrons-expand dnd_icon"
          data-draggable="true"
        />
        <div
          class="choice"
        >
          <div
            class="small "
          >
            <temba-select
              data-testid="temba_select_operator"
              errors="[]"
              name="operator"
              namekey="verboseName"
              valuekey="type"
              values="[{\\"type\\":\\"has_number_between\\",\\"verboseName\\":\\"has a number between\\",\\"operands\\":2}]"
            />
          </div>
        </div>
        <div
          class="multi_operand"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                type="text"
                value="1"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <span
            class="divider"
            data-draggable="true"
          >
            and
          </span>
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="arguments"
                type="text"
                value="100"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="categorize_as"
          data-draggable="true"
        >
          categorize as
        </div>
        <div
          class="category"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="36"
                name="Exit Name"
                type="text"
                value="1 - 100"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="fe-x remove_icon"
          data-testid="remove-case-763e4844-3e1b-407a-a1b5-5fdfcd308b41"
        />
      </div>
    </div>
  </div>
</body>
`;
